pulp_puppet_plugins/pulp_puppet/plugins/importers/forge.py

```diff
@@ -311,14 +315,32 @@ def _add_new_module(self, downloader, module):
         finally:
             downloader.cleanup_module(module)
 
-    def _resolve_new_units(self, existing_unit_keys, metadata_unit_keys):
+    def _resolve_new_units(self, existing, wanted):
         """
-        Returns a list of metadata keys that are new to the repository.
+        Decide what units are needed to be downloaded.
 
-        :return: list of unit keys; empty list if none are new
-        :rtype:  list
+        Filter out units which are already in a repository,
+        associate units which are already downloaded,
+
+        :param existing: units which are already in a repository
+        :type existing: list of unit keys as namedtuples
+        :param wanted: units which should be imported into a repository
+        :type wanted: list of unit keys as namedtuples
+
+        :return: list of unit keys to download; empty list if all units are already downloaded
+        :rtype:  list of unit keys as namedtuples
         """
-        return list(set(metadata_unit_keys) - set(existing_unit_keys))
+        model = plugin_api.get_unit_model_by_id(constants.TYPE_PUPPET_MODULE)
+        unit_generator = (model(**unit_tuple._asdict()) for unit_tuple in wanted)
+        still_wanted = set(wanted)
+        for unit in units_controller.find_units(unit_generator):
+            file_exists = unit._storage_path is not None and os.path.isfile(unit._storage_path)
+            if file_exists:
+                if unit.unit_key_as_named_tuple not in existing:
```

Comment:
it is purely matter of personal preference but i would move line 341 to right after condition on line 338 happens, it is more readable imho